data_safe_haven/infrastructure/stack_manager.py

```diff
@@ -1,24 +1,113 @@
 """Deploy with Pulumi"""
 import logging
-import os
 import pathlib
 import shutil
 import subprocess
 import time
 from contextlib import suppress
 from importlib import metadata
+from shutil import which
 from typing import Any
 
 from pulumi import automation
+import typer
 
 from data_safe_haven.config import Config
 from data_safe_haven.exceptions import DataSafeHavenAzureError, DataSafeHavenPulumiError
-from data_safe_haven.external import AzureApi, AzureCli
+from data_safe_haven.external import AzureApi
 from data_safe_haven.functions import replace_separators
 from data_safe_haven.infrastructure.stacks import DeclarativeSHM, DeclarativeSRE
 from data_safe_haven.utility import LoggingSingleton
 
 
+def handle_pulumi_login(config) -> None:
+    account = PulumiAccount(config)
+
+    if not account.confirm():
+        account.login()
+        if not account.confirm():
+            raise typer.Exit()
+
+
+def pulumi_env(config: Config) -> dict[str, Any]:
+    azure_api = AzureApi(config.subscription_name)
+    backend_storage_account_keys = azure_api.get_storage_account_keys(
+        config.backend.resource_group_name,
+        config.backend.storage_account_name,
+    )
+    return {
+        "AZURE_STORAGE_ACCOUNT": config.backend.storage_account_name,
+        "AZURE_STORAGE_KEY": str(backend_storage_account_keys[0].value),
+        "AZURE_KEYVAULT_AUTH_VIA_CLI": "true",
+    }
```

Comment:
Is this useful anywhere outside PulumiAccount? Feels like it could be part of that class.

Reply:
I'm generally a fan of Smalltalk-style classes which ask each other things. So if we think that the `pulumi_env` is conceptually *part of* the `PulumiAccount` then the way to get hold of it should be to ask `PulumiAccount` "what is the value of `pulumi_env`" through a method like `env`.